### PR TITLE
Bug 1945091: Refactor openshift-tests a bit, add more network-test-autodetecting support

### PR DESCRIPTION
**Pull request by danwinship**
Currently we are unconditionally skipping all single-stack-IPv6, dual-stack, and SCTP tests, regardless of cluster configuration. This is no good.

This PR extends the existing network-plugin-detection code to also detect network IP configuration and automatically skip IPv4-only, IPv6-only, and dual-stack-only tests as appropriate.

It also kind of adds support for detecting and skipping SCTP tests, except we can't really easily detect whether SCTP is available in the cluster at this point, so in reality it only works if you manually override the SCTP detection via `--provider`/`TEST_PROVIDER`. But we might be able to improve on that in the future. (And being able to unskip the SCTP tests at all is an improvement over the current situation.)

This will require a followup PR to openshift/kubernetes to stop skipping these tags. Thus, for the moment, this PR should have no effect on the tests that actually get run.

Also, until https://issues.redhat.com/browse/KNIDEPLOY-4090 is fixed, the `e2e-metal-ipi-ovn-ipv6` and `e2e-metal-ipi-ovn-dualstack` suites won't actually run any of the "fun" test cases anyway...

### CODE REVIEW

**Comment by marun** [comment]
/lgtm

This looks fine to me, and I appreciate the testing you've added. 

Any reason not to unskip in o/k first and vendor that change as part of this PR to get some test signal? Any change to o/k will have to be vendored to origin to take effect.

**Comment by danwinship** [comment]
> Any reason not to unskip in o/k first and vendor that change as part of this PR to get some test signal? Any change to o/k will have to be vendored to origin to take effect.

Filed that PR now, https://github.com/openshift/kubernetes/pull/565. But that shouldn't merge until after this does since if it got vendored before this PR merged it would cause tests to be unskipped that shouldn't be unskipped (eg running dual-stack tests in single-stack clusters).

But also, this PR is mostly just a prereq; we still have to fix KNIDEPLOY-4090 to get the IPv6/dual-stack tests to actually run, and we still have to create a new job (or modify some existing one) to get the SCTP tests to run.

**Comment by danwinship** [comment]
/hold
testing with updated annotations; should not actually affect what runs, in the end

**Comment by danwinship** [comment]
e2e-gcp runs [with](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/25818/pull-ci-openshift-origin-master-e2e-gcp/1359600677319348224) and [without](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/25818/pull-ci-openshift-origin-master-e2e-gcp/1359561169513222144) the vendored kube change are identical (other than having different flakes). (The different test count is just because they had different numbers of flakes and so had different numbers of retries. Sorting and diffing the junit output shows no other differences.)

File: test/extended/util/cluster/cluster.go [hunk 1]
```diff
@@ -174,12 +196,29 @@ func SetTestConfig(platformStatus *configv1.PlatformStatus, masters, nonMasters
...
 		if c.NetworkPluginMode != "" {
 			skips = append(skips, fmt.Sprintf("[Skipped:Network/%s/%s]", c.NetworkPlugin, c.NetworkPluginMode))
 		}
 	}
+
+	if !c.HasIPv4 {
+		skips = append(skips, "[Feature:Networking-IPv4]")
+	}
+	if !c.HasIPv6 {
+		skips = append(skips, "[Feature:Networking-IPv6]")
```
Comment (aojea) [inline test/extended/util/cluster/cluster.go:211]: just FYI these tags are not longer upstream, tests must work on both IP families, there is only one test remaining that I couldn't migrate

```
grep -r Networking  test | grep Fea
test/e2e/network/networking_perf.go:var _ = SIGDescribe("Networking IPerf2 [Feature:Networking-Performance]", func() {
test/e2e/network/networking.go: ginkgo.It("should provide Internet connection for containers [Feature:Networkin-IPv4]", func() {
test/e2e/network/networking.go: ginkgo.It("should provide Internet connection for containers [Feature:Networkin-IPv6][Experimental][LinuxOnly]", func() {
```

Comment (aojea) [reply test/extended/util/cluster/cluster.go:211]: heh, I just realise the tag is wrong Networkin/Networking

File: test/extended/util/cluster/cluster.go [hunk 2]
```diff
@@ -170,3 +171,22 @@ func SetTestConfig(platformStatus *configv1.PlatformStatus, masters, nonMasters
 	testNonMasters = nonMasters
 	testNetworkSpec = networkSpec
 }
+
+// GetMatchFn returns a function that tests if a named function should be run based on
+// the cluster configuration
+func (c *ClusterConfiguration) GetMatchFn() func(string) bool {
```
Comment (smarterclayton) [inline test/extended/util/cluster/cluster.go:196]: Call this MatchFn()

File: cmd/openshift-tests/provider.go [hunk 3]
```diff
@@ -95,49 +95,16 @@ func decodeProvider(provider string, dryRun, discover bool) (*exutilcluster.Clus
 		if len(providerInfo.Type) == 0 {
 			return nil, fmt.Errorf("provider must be a JSON object with the 'type' key")
 		}
-		var cloudConfig e2e.CloudConfig
```
Comment (smarterclayton) [inline cmd/openshift-tests/provider.go:124]: Doing this would mean we have to be backwards compatible forever in upstream test context (the config for a step is ignorant of the version of openshift it’s testing).  If upstream changes the struct in a non backwards compatible way the rebase will require all the test steps/config in CI to be forked).

I’m not opposed to the change in theory but I’d want some evidence we have a way that is not “openshift-tests api surface to CI changes” before moving forward.

**Comment by smarterclayton** [comment]
A few concerns on individual commits.

File: cmd/openshift-tests/provider.go (hunk 3 above)
Comment (danwinship) [reply cmd/openshift-tests/provider.go:124]: No, it's the other way around; _currently_ we expose the upstream API in our API. After this PR we do not.

That is, before this commit we do:
  - unmarshal `provider` into an `e2e.CloudConfig`
  - copy the `e2e.CloudConfig` into an `exutilcluster.ClusterConfig` element-by-element
  - later, copy the `exutilcluster.ClusterConfig` _back_ into an `e2e.CloudConfig`, element-by-element, for the tests to use (https://github.com/openshift/origin/blob/ff70d2d5/cmd/openshift-tests/provider.go#L31)

after this PR we do:
  - unmarshal `provider` into an `exutilcluster.ClusterConfig`
  - later, copy the `exutilcluster.ClusterConfig` into an `e2e.CloudConfig`, element-by-element

Comment (smarterclayton) [reply cmd/openshift-tests/provider.go:127]: We're backward compatible on the json format all the way back to 4.1 still, right?

Comment (danwinship) [reply cmd/openshift-tests/provider.go:127]: Yes. The JSON-parsing unit test is added in the first commit, and none of the input values of those tests is changed after that. (The _output_ values change slightly to reflect two bugfixes: we now correctly set the default vales of NumNodes and MultiMaster even on bare metal; and we no longer ignore the MultiMaster and MultiZone fields in the input JSON if it's trying to override them.)

If you want to add additional examples from openshift/release to the unit tests I can do that

**Comment by smarterclayton** [comment]
I'm just waiting for bens thing to merge before i go back and do another pass, will try to accelerate review here

**Comment by danwinship** [comment]
> I'm just waiting for bens thing to merge before i go back and do another pass, will try to accelerate review here

By "bens thing" do you mean Stephen Benjamin's PR #26019 ? That's on /hold right now...

**Comment by stbenjam** [comment]
It’ll be unheld today. The new annotations messed up our job filters and should be fixed.

**Comment by danwinship** [comment]
rebased (including adding tests for disconnected mode to the unit test)

File: test/extended/util/cloud/cloud.go [hunk 4]
```diff
@@ -43,65 +45,76 @@ func (c *ClusterConfiguration) ToJSONString() string {
...
-	}
-	operatorClient, err := operatorclient.NewForConfig(clientConfig)
-	if err != nil {
-		return nil, err
-	}
+// DiscoverConfig uses the cluster to setup the cloud provider config.
+func DiscoverConfig() (*ClusterConfiguration, error) {
+	p, masters, nonMasters, networkSpec := testPlatformStatus, testMasters, testNonMasters, testNetworkSpec
+	if p == nil {
+		clientConfig, err := e2e.LoadConfig(true)
```
Comment (smarterclayton) [inline test/extended/util/cloud/cloud.go]: I explicitly did not want this package depending on e2e framework (which brings in the world).  Please move this back out.

Comment (smarterclayton) [reply test/extended/util/cloud/cloud.go]: Our goal is to wean ourselves off of upstream e2e for all the core infra.

File: test/extended/util/cloud/cloud.go [hunk 5]
```diff
@@ -43,65 +45,76 @@ func (c *ClusterConfiguration) ToJSONString() string {
...
-	if err != nil {
-		return nil, err
-	}
-	operatorClient, err := operatorclient.NewForConfig(clientConfig)
-	if err != nil {
-		return nil, err
-	}
+// DiscoverConfig uses the cluster to setup the cloud provider config.
+func DiscoverConfig() (*ClusterConfiguration, error) {
+	p, masters, nonMasters, networkSpec := testPlatformStatus, testMasters, testNonMasters, testNetworkSpec
```
Comment (smarterclayton) [inline test/extended/util/cloud/cloud.go]: No, please no. No globals.  If you want to preserve stuff, keep it in a struct you return from this package.

File: cmd/openshift-tests/openshift-tests.go [hunk 6]
```diff
@@ -29,7 +29,7 @@ import (
 	testginkgo "github.com/openshift/origin/pkg/test/ginkgo"
 	"github.com/openshift/origin/pkg/version"
 	exutil "github.com/openshift/origin/test/extended/util"
-	"github.com/openshift/origin/test/extended/util/cloud"
+	"github.com/openshift/origin/test/extended/util/cluster"
```
Comment (smarterclayton) [inline cmd/openshift-tests/openshift-tests.go:32]: This rename is fine, you could also pull this out into its own PR to reduce churn in this one

File: cmd/openshift-tests/e2e.go [hunk 7]
```diff
@@ -401,13 +401,13 @@ func suiteWithInitializedProviderPreSuite(opt *runOptions) error {
 }
 
 // suiteWithProviderPreSuite ensures that the suite filters out tests from providers
-// that aren't relevant (see getProviderMatchFn) by loading the provider info from the
-// cluster or flags.
+// that aren't relevant (see exutilcluster.ClusterConfig.MatchFn) by loading the
```
Comment (smarterclayton) [inline cmd/openshift-tests/e2e.go:404]: This refactor is fine, you could also pull this out into its own PR to reduce churn in this one

File: test/extended/util/cluster/cluster.go [hunk 8]
```diff
@@ -104,12 +104,6 @@ func DiscoverConfig() (*ClusterConfiguration, error) {
 		networkPluginIDs = append(networkPluginIDs, string(networkSpec.DefaultNetwork.Type)+"/"+string(networkSpec.DefaultNetwork.OpenShiftSDNConfig.Mode))
 	}
 
-	if p.Type == configv1.NonePlatformType {
```
Comment (smarterclayton) [inline test/extended/util/cluster/cluster.go]: Also non controversial, could split with the other commits and get it merged faster

**Comment by smarterclayton** [comment]
Just the two changes needed:

1. remove use of globals introduced in first commit
2. don't depend on e2e.framework within our util cluster/cloud package, pass config in.

**Comment by danwinship** [comment]
/retest
e2e-gcp-upgrade appears to be mostly failing for everyone
